### ISSUE TITLE
Replace Ember with Angular 

### ISSUE THREAD

**Issue by chrislaughlin** [cam-technologies/time-booker#52]
I know we have not been active in a while but I would like to see this come to some sort of working version. I have been playing with the idea in my head for a while now and I feel that the fact that we all have some level of experience with Angular JS could leverage some of the learning curve that has stumped the UI development to date. 

The main reason for choosing the MEEN stack was to give us a chance to learn Ember in a working however jumping in head first has thrown up some challenges testing to name but a few. The view I seem to get from watching the different JavaScript sub reddit's is that Angular is not only building up steam but becoming a strong player. We have had issues with testing Ember and not really getting unit tests that were easy to write, compared to Angular's strong testing background.   

The main goal of the project is to learn and home JavaScript skills this will not changes as with the existing Angular projects we have used we had only touched on the surface and in most cases went against most of the known best practices, we still have the chance to build a client that will follow industry standards as well build upon strong framework skills we already have. 

This will also enable us to benefit from the MEAN stack and use the newest features of Angular and integrated with the latests frameworks. This change would require no alterations to the server side as we currently have a boiler plate for the client.   

What do you guys think?

**Comment by madole** [comment]
I've actually been thinking about this too as I've started to use angular. I think it's a pretty easy framework to work with. Much lower learning curve and you guys already have a wealth of experience with it. 

We could do the MEAN.io stack, and there are yeoman generators out there that i've been using that make it really easy to add services/directives/controllers/routes etc. 

I think this could be a change that could get us back on track!

**Comment by chrislaughlin** [comment]
Great man, I would even be tempted to go the whole hog and throw away the current repo and drop in a MEAN stack from a generator as our npm packages have grown out of control and the application has taken on a life of its own.

**Comment by martinmicunda** [comment]
oooo so we are back on the track :) I agree the Ember is not good choice to the future as what I know there are only few large companies that are using Ember.js like Zendesk, Intercom.... Personally the problem what I see with Ember is break app to small pieces (modules) as Ember has really MVC structure similer to Ruby on Rails. Regarding to MEAN.io or other generators I wouldn't use them as most of them in my opinion don't follow best practices for developing Angular app see John Pappa [angularjs-styleguide](https://github.com/johnpapa/angularjs-styleguide) (Note: I still think his style guide needs to be improve specially the part how to split app by feature and how to separate business logic.).

Also if we are choosing Angular I would start writing code in ECMAScript 6 so migration to Angular 2 would be easy and we can start already using some of the Angular 2 modules that are written in ECMAScript 6. 

I would switch from Express.js to Koa.js and also start looking to Ionic framework for Angular mobile version.

And last think if you have time you can check my [employee-scheduling](https://github.com/martinmicunda/employee-scheduling) that I been working on..

I am going to Paris for Angular ngEurope conference next week so I hope that will say something more about Angular 2.

**Comment by chrislaughlin** [comment]
I agree on some points @martinmicunda times have changed with YOMAN and some of the generators now follow the best practices when it comes to file and folders as YOMAN would only really be used for adding new components its not a big deal if they are not 100% as we will be fixing them up before committing. I'm still a little low in my scope of the Angular 2 dev with ECMAScript 6. Will this be supported by most browsers (Chrome, FF and IE 10+)?

What are the advantages of Koa over express? Iv not used Koa but have only really used the basics in express so it might not be a problem.     

I found this to be another good style guide http://toddmotto.com/opinionated-angular-js-styleguide-for-teams/?utm_content=buffer88782&utm_medium=social&utm_source=twitter.com&utm_campaign=buffer

I think that the MEAN.io solution could be a good building block as it has the auth work already done which we did struggle with. I really want o stay away from the the build process we currently have as it does tone of stuff that I don't know or understand and has been very unstable. 

We need to take it slow this time as things got out of hand before, when adding new modules and build processes we need to take them one at a time and all agree before added them. 

Hope you have a good time in Paris, I'm still waiting on my stickers from node conf EU

**Comment by martinmicunda** [comment]
@chrislaughlin JavaScript 6 will be released in March 2015 however it will take some time when the features will be implemented to all browser but u can use [traceur-compiler](https://github.com/google/traceur-compiler) which compile your code from JS 6 to JS 5. Koa is written in JS 6 and it's try solve some pain issues with JS like error handling etc...(I also haven't been using it but would be good to check that project). Regarding to MEAN.io the main person left the project and create new one calling [Mean.js](http://meanjs.org/)

**Comment by chrislaughlin** [comment]
I would be worries about taking on es6 if we have limited support would it
be possible to write in es5 then upgrade. Adding another compiler to the
mix might over complicate things.

On Wednesday, October 15, 2014, Martin Micunda notifications@github.com
wrote:

> @chrislaughlin https://github.com/chrislaughlin JavaScript 6 will be
> released in March 2015 however it will take some time when the features
> will be implemented to all browser but u can use traceur-compiler
> https://github.com/google/traceur-compiler which compile your code from
> JS 6 to JS 5. Koa is written in JS 6 and it's try solve some pain issues
> with JS like error handling etc...(I also haven't been using it but would
> be good to check that project). Regarding to MEAN.io the main person left
> the project and create new one calling Mean.js http://meanjs.org/
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cam-technologies/time-booker/issues/52#issuecomment-59284694
> .

## 

Sent from Gmail Mobile

**Comment by madole** [comment]
I'm for ES6, It's the future of JS and most things are already supported by Chrome and Firefox...

As for Koa, I'm happy to use it too. V8 doesn't do ES6 yet but you can run node in --harmony mode and it will enable the ES6 features.

**Comment by chrislaughlin** [comment]
I'm afraid that we will end up wasting time getting ES6 working and
compiling for the browser and not actually focusing on development. Chrome
does support some of it now but it will be at least a year for others to
take it on. More running under harmony is not stable and I'm not see
totally supported for production.

If we focus on building an application on the strong standards that are out
now we have a base that we can upgrade.

On Wednesday, October 15, 2014, Andrew McDowell notifications@github.com
wrote:

> I'm for ES6, It's the future of JS and most things are already supported
> by Chrome and Firefox...
> 
> As for Koa, I'm happy to use it too. V8 doesn't do ES6 yet but you can run
> node in -- mode and it will enable the ES6 features.
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cam-technologies/time-booker/issues/52#issuecomment-59285513
> .

## 

Sent from Gmail Mobile

**Comment by madole** [comment]
I see where you're coming from but let's live in the bleeding edge. 

Traceur is used by lots of companies in production. Its been about for a while pokyfilling ES6. Node harmony doesn't need to be 100% stable for a Dev iteration. But I think the instabilities are played up a bit. I don't think it's that bad. 

If we code got ES6, when V8 adopts it, we are flying with an up to date codebase. 

We're not dependent on this for income which is why you'd play it safe in the first place. Let's live a little!

**Comment by chrislaughlin** [comment]
Okay I'm happy to go for es6 as long as we keep the development cycle
simple and fast, battling against compilers will only put us of working on
it.

On Wednesday, October 15, 2014, Andrew McDowell notifications@github.com
wrote:

> I see where you're coming from but let's live in the bleeding edge.
> 
> Traceur is used by lots of companies in production. Its been about for a
> while pokyfilling ES6. Node harmony doesn't need to be 100% stable for a
> Dev iteration. But I think the instabilities are played up a bit. I don't
> think it's that bad.
> 
> If we code got ES6, when V8 adopts it, we are flying with an up to date
> codebase.
> 
> We're not dependent on this for income which is why you'd play it safe in
> the first place. Let's live a little!
> 
> —
> Reply to this email directly or view it on GitHub
> https://github.com/cam-technologies/time-booker/issues/52#issuecomment-59286501
> .

## 

Sent from Gmail Mobile

**Comment by madole** [comment]
Yeah, I really think our build process needs to grow organically, extended as and when we need it. 

As little as possible to make the code build first then as we need features, we build them in.

**Comment by chrislaughlin** [comment]
Plus one on that @madole 

So what will we take at the next step and remember baby steps :)

**Comment by chrislaughlin** [comment]
We can k

**Comment by chrislaughlin** [comment]
Dam github mobile site lol

**Comment by chrislaughlin** [comment]
Over the weekend I'll clean down the repo and update the read me 

A fresh start to kick off the development?

**Comment by madole** [comment]
Yeah sounds good! Let's clear our minds and start again! We can do this!
